src/test/java/com/google/firebase/ThreadManagerTest.java

```diff
@@ -0,0 +1,227 @@
+    FirebaseApp app = FirebaseApp.initializeApp(buildOptions(threadManager));
+
+    // Initializing an app doesn't initialize any threading resources.
+    assertEquals(0, threadManager.events.size());
+
+    // Initializing auth should not initialize threading resources.
+    FirebaseAuth auth = FirebaseAuth.getInstance(app);
+    assertEquals(0, threadManager.events.size());
+
+    // Calling an async method on auth should initialize threading resources.
+    try {
+      auth.verifyIdTokenAsync("foo").get();
+    } catch (InterruptedException | ExecutionException ignored) {
+      // ignored
+    }
+    assertEquals(1, threadManager.events.size());
+    Event event = threadManager.events.get(0);
+    assertEquals(Event.TYPE_GET_EXECUTOR, event.type);
+    assertEquals(app, event.arguments[0]);
+
+    // Calling a method again should not re-initialize threading resources.
+    try {
+      auth.verifyIdTokenAsync("foo").get();
+    } catch (InterruptedException | ExecutionException ignored) {
+      // ignored
+    }
+    assertEquals(1, threadManager.events.size());
+
+    // Should not re-initialize threading resources.
+    TestOnlyImplFirebaseTrampolines.forceThreadManagerInit(app);
+    assertEquals(1, threadManager.events.size());
+
+    // Deleting app should tear down threading resources.
+    app.delete();
+    assertEquals(2, threadManager.events.size());
+    event = threadManager.events.get(1);
+    assertEquals(Event.TYPE_RELEASE_EXECUTOR, event.type);
+    assertEquals(app, event.arguments[0]);
+  }
+
+  private FirebaseOptions buildOptions(ThreadManager threadManager) {
+    return new FirebaseOptions.Builder()
+        .setCredentials(new MockGoogleCredentials())
+        .setProjectId("mock-project-id")
+        .setThreadManager(threadManager)
+        .build();
+  }
+
+  public static class MockThreadManager extends ThreadManager {
+
+    private final List<Event> events = new ArrayList<>();
+
+    @Override
+    protected ScheduledExecutorService getExecutor(@NonNull FirebaseApp app) {
+      events.add(new Event(Event.TYPE_GET_EXECUTOR, app));
+      return Executors.newSingleThreadScheduledExecutor();
+    }
+
+    @Override
+    protected void releaseExecutor(@NonNull FirebaseApp app,
+        @NonNull ScheduledExecutorService executor) {
+      events.add(new Event(Event.TYPE_RELEASE_EXECUTOR, app, executor));
+      executor.shutdownNow();
+    }
+
+    @Override
+    protected ThreadFactory getThreadFactory() {
+      events.add(new Event(Event.TYPE_GET_THREAD_FACTORY));
+      return Executors.defaultThreadFactory();
+    }
+  }
+  
+  private static class Event {
+    private static final int TYPE_GET_EXECUTOR = 100;
+    private static final int TYPE_RELEASE_EXECUTOR = 101;
+    private static final int TYPE_GET_THREAD_FACTORY = 102;
+
+    private final int type;
+    private final Object[] arguments;
+
+    public Event(int type, Object... arguments) {
```

Comment:
I would prefer if you explicitly called out the arguments.

Reply:
Done